### PR TITLE
Add functions for after-change style of CSS Transition 

### PR DESCRIPTION
**Pull request by hiikezoe**
This is a PR of https://bugzilla.mozilla.org/show_bug.cgi?id=1346663

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [X] These changes do not require tests because it's for stylo

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16111)

### CODE REVIEW

**Comment by hiikezoe** [comment]
All patches have been reviewed by @emilio.

File: components/style/matching.rs [hunk 1]
```diff
@@ -605,6 +605,33 @@ trait PrivateMatchMethods: TElement {
...
+            context.shared.stylist.rule_tree.remove_transition_rule_if_applicable(rule_node);
+        if without_transition_rules == *rule_node {
+            // Note that unwrapping here is fine, because the style is
+            // only incomplete during the styling process.
+            return style.values.as_ref().unwrap().clone();
+        }
+
+        let mut cascade_flags = CascadeFlags::empty();
+        if self.skip_root_and_item_based_display_fixup() {
+          cascade_flags.insert(SKIP_ROOT_AND_ITEM_BASED_DISPLAY_FIXUP)
```
Comment (emilio) [inline components/style/matching.rs]: nit: indentation.

**Review by emilio — APPROVED** [review]

File: components/style/rule_tree/mod.rs [hunk 2]
```diff
@@ -728,6 +759,13 @@ impl StrongRuleNode {
             self.gc();
         }
     }
+
+    /// Returns true if there is either animation or transition level rule.
+    pub fn has_animation_or_transition_rules(&self) -> bool {
+        self.self_and_ancestors().take_while(|node| node.cascade_level() >= CascadeLevel::Animations)
+                                 .any(|node| node.cascade_level() == CascadeLevel::Animations ||
```
Comment (emilio) [inline components/style/rule_tree/mod.rs]: consider indenting the `take_while` and `any` with the `self_and_ancestors`. Also, I believe you can use the `matches!` macro.

```rust
self.self_and_ancestors()
    .take_while(|node| node.cascade_level() >= CascadeLevel::Animations)
    .any(|node| matches!(node.cascade_level(), CascadeLevel::Animations | CascadeLevel::Transitions))
```

**Review by emilio — APPROVED** [review]
Feel free to r=me with the indentation nit addressed (and the other one if you want).

Thanks for doing this!

File: components/style/rule_tree/mod.rs (hunk 2 above)
Comment (hiikezoe) [inline components/style/rule_tree/mod.rs]: Thanks! I did not know the matches! macro!

**Pull request merged**